### PR TITLE
Htk sharing/utilities refactor/formatting/fixes

### PR DESCRIPTION
**Pull request by HodgsonSDAS**
After the big sharing service update (https://github.com/Microsoft/HoloToolkit-Unity/pull/459) there have been quite a few issues.  This PR aims to fix those.

Updated Sharing Readme. [Preview Here](https://github.com/HodgsonSDAS/HoloToolkit-Unity/blob/HTK-SharingRefactor/Assets/HoloToolkit/Sharing/README.md)
Updated Utilities Readme. [Preview Here](https://github.com/HodgsonSDAS/HoloToolkit-Unity/blob/HTK-SharingRefactor/Assets/HoloToolkit/Utilities/README.md)

Resolves https://github.com/Microsoft/HoloToolkit-Unity/issues/483 : Fixes `Sharing.prefab` warning.
Resolves https://github.com/Microsoft/HoloToolkit-Unity/issues/480 : Adds better docs for sharing services, and a new test script named `UserNotifications.cs`.
Resolves https://github.com/Microsoft/HoloToolkit-Unity/issues/491 : Fixes null reference when sharing stage `AutoDiscoverServer == true`.

- Refactored a few classes that probably belonged in Utilities folder.
- Organized Sharing/Tests folder.
- Fixed test scenes in sharing folder
    - Missing `HoloLensCamera.prefab`
    - `SharingTest.scene` did not have server address set to localhost
- Cleaned up formatting and code in various files.
- Removed redundant code.
- Fixed most object boxing issues.

I also found two redundant scripts but wasn't sure what to do with them `Assets/Utilities/Scripts/InterpolatedValues`:

- QuaternionInterpolated.cs
- Vector3Interpolated.cs

Both seem to be similar in function to the existing scripts in this folder.

### CODE REVIEW

File: Assets/HoloToolkit/Utilities/README.md [hunk 1]
```diff
@@ -14,10 +15,101 @@ A drop in direction indicator that stays in the users view at all times.
 A quad based pointer to be used with the HeadsUpDirectionIndicator prefab to create an out of box direction indicator.
 
 ### [Scripts](Scripts)
+---
```
Comment (matthejo) [inline Assets/HoloToolkit/Utilities/README.md]: Thanks for helping fill out this documentation!

File: Assets/HoloToolkit/Utilities/Scripts/WorldAnchorManager.cs [hunk 2]
```diff
@@ -170,7 +170,7 @@ private void CreateAnchor(GameObject gameObjectToAnchor, string anchorName)
             }
             else
             {
-                // Othertimes we must wait for the 
+                // Othertimes we must wait for the
```
Comment (matthejo) [inline Assets/HoloToolkit/Utilities/Scripts/WorldAnchorManager.cs]: Nit: not your change at all, but this is probably meant to say "Other times we must wait for the tracking system to locate the world" or something along those lines.

File: Assets/HoloToolkit/Sharing/Scripts/SessionUsersTracker.cs [hunk 3]
```diff
@@ -115,9 +118,18 @@ private void OnUserJoinedSession(Session session, User user)
...
-				CurrentUsers.RemoveAll(x => x.GetID() == user.GetID()); // in case there was an old user with the same ID
-				CurrentUsers.Add(user);
+                //Debug.LogFormat("User {0} joined current session.", user.GetName());
+
+                // Remove any old users with the same ID
+                for (int i = 0; i < CurrentUsers.Count; i++)
+                {
+                    if (CurrentUsers[i].GetID() == user.GetID())
+                    {
+                        CurrentUsers.Remove(user);
```
Comment (HodgsonSDAS) [inline Assets/HoloToolkit/Sharing/Scripts/SessionUsersTracker.cs]: yeah I noticed this testing. I'll revert this.

File: Assets/HoloToolkit/Sharing/Scripts/Spawning/SyncSpawnArray.cs [hunk 4]
```diff
@@ -33,16 +27,17 @@ protected override T CreateObject(ObjectElement objectElement)
...
 
-            System.Object createdObject = Activator.CreateInstance(typeToInstantiate);
+            object createdObject = Activator.CreateInstance(typeToInstantiate);
 
-            T spawnedDataModel = createdObject as T;
+            T spawnedDataModel = (T)createdObject;
             spawnedDataModel.Element = objectElement;
             spawnedDataModel.FieldName = objectElement.GetName();
-            spawnedDataModel.Owner = this.Owner; // TODO: this should really come from objectElement.GetOwnerID().  It not safe to assume that this class and its child have the same owner
+            // TODO: this should not query SharingStage, but instead query the underlying session layer
```
Comment (HodgsonSDAS) [inline Assets/HoloToolkit/Sharing/Scripts/Spawning/SyncSpawnArray.cs]: There's a [place in SyncArray that](https://github.com/Microsoft/HoloToolkit-Unity/blob/master/Assets/HoloToolkit/Sharing/Scripts/SyncModel/SyncArray.cs#L76) that looked very similar and the TODO is from copying this over.

I thought getting the owner this way made more sense than just setting this.Owner.

File: Assets/HoloToolkit/Sharing/Scripts/SyncModel/SyncArray.cs [hunk 5]
```diff
@@ -37,19 +37,16 @@ public class SyncArray<T> : SyncObject, IEnumerable<T> where T : SyncObject, new
         /// Type of objects in the array.
         /// This is cached so that we don't have to call typeof(T) more than once.
         /// </summary>
-        protected Type arrayType;
+        protected Type ArrayType;
```
Comment (HodgsonSDAS) [inline Assets/HoloToolkit/Sharing/Scripts/SyncModel/SyncArray.cs]: agreed.

File: Assets/HoloToolkit/Sharing/Scripts/SyncModel/SyncArray.cs [hunk 6]
```diff
@@ -230,13 +222,13 @@ private T AddObject(ObjectElement existingElement)
...
             // Update internal map
-            this.dataArray[id] = newObject;
+            dataArray[id] = newObject;
 
             // If it's local, make sure to initialize it so it can be used immediately.
-            if (isLocal)
-            {
-                newObject.InitializeLocal(this.Element);
-            }
+            //if (isLocal)
```
Comment (HodgsonSDAS) [inline Assets/HoloToolkit/Sharing/Scripts/SyncModel/SyncArray.cs]: Sounds good.

File: Assets/HoloToolkit/Sharing/Scripts/SyncModel/SyncQuaternion.cs [hunk 7]
```diff
@@ -18,26 +18,25 @@ public class SyncQuaternion : SyncObject
         [SyncData] private SyncFloat z = null;
         [SyncData] private SyncFloat w = null;
 
+#if UNITY_EDITOR
         public override object RawValue
         {
-            get { return new Quaternion(this.x.Value, this.y.Value, this.z.Value, this.w.Value); }
+            get { return new Quaternion(x.Value, y.Value, z.Value, w.Value); }
```
Comment (HodgsonSDAS) [inline Assets/HoloToolkit/Sharing/Scripts/SyncModel/SyncQuaternion.cs]: I'll do the same for SyncVector3 as well

File: Assets/HoloToolkit/Utilities/README.md (hunk 1 above)
Comment (HodgsonSDAS) [inline Assets/HoloToolkit/Utilities/README.md]: No worries!

File: Assets/HoloToolkit/Utilities/Scripts/WorldAnchorManager.cs (hunk 2 above)
Comment (HodgsonSDAS) [inline Assets/HoloToolkit/Utilities/Scripts/WorldAnchorManager.cs]: Fixed.

File: Assets/HoloToolkit/Sharing/README.md [hunk 8]
```diff
@@ -1,18 +1,117 @@
 ## [Sharing]()
-Sharing and networking components for rapid prototyping in Unity for building shared experiences.
+The HoloToolkit.Sharing library allows applications to span multiple devices, and enables holographic collaboration.
```
Comment (NeerajW) [inline Assets/HoloToolkit/Sharing/README.md]: This is fab :)

**Pull request merged**